### PR TITLE
Make src gen for property setters consistent with reflection

### PR DESCRIPTION
**Pull request by steveharter**
Address https://github.com/dotnet/runtime/issues/86333 (RC2 candidate as well)
- Call the setter for value type properties (with the default value for the property type) if the value is not in the config, unless it is during a Bind (not "Get"). This requires a adding a `bool` arg to the `BindCore()` method.
- Always check for a `null` value or missing section for properties and if null\missing (this was not done in all cases previously) then don't call the setter with the `null`. This addresses https://github.com/dotnet/runtime/issues/91380.

Note that this PR matches known semantics regarding the above with the goal of maintaining compat with reflection; this is the case even for non-intuitive cases including:
- Only value type properties get their setter called; not reference types or `Nullable<T>`.
- Also, properties on objects that are nested objects or those added to a collection are also not defaulted through the setter (only properties on root objects are).

We could create an issue for these non-intuitive cases, but would need to consider the benefit:risk including breaking changes.

### CODE REVIEW

**Comment by tarekgh** [comment]
@steveharter there are some failures that seem related to the change. Could you please have a look?

File: src/libraries/Microsoft.Extensions.Configuration.Binder/tests/SourceGenerationTests/Baselines/OptionsBuilder/Bind_T_BinderOptions.generated.txt
Comment (layomia) [inline src/libraries/Microsoft.Extensions.Configuration.Binder/tests/SourceGenerationTests/Baselines/OptionsBuilder/Bind_T_BinderOptions.generated.txt:113]:
```suggestion
                instance.MyString = value1;
```

File: src/libraries/Microsoft.Extensions.Configuration.Binder/tests/SourceGenerationTests/Baselines/OptionsBuilder/Bind_T_BinderOptions.generated.txt
Comment (layomia) [inline src/libraries/Microsoft.Extensions.Configuration.Binder/tests/SourceGenerationTests/Baselines/OptionsBuilder/Bind_T_BinderOptions.generated.txt:122]: Perhaps make this a conditional assignment to reduce LOC?

File: src/libraries/Microsoft.Extensions.Configuration.Binder/tests/SourceGenerationTests/Baselines/OptionsBuilder/Bind_T.generated.txt
Comment (steveharter) [reply src/libraries/Microsoft.Extensions.Configuration.Binder/tests/SourceGenerationTests/Baselines/OptionsBuilder/Bind_T.generated.txt:120]: The reflection appears to only call the setter for value types and not `Nullable<T>` -- basically something that can never be `null`.

File: src/libraries/Microsoft.Extensions.Configuration.Binder/tests/SourceGenerationTests/Baselines/OptionsBuilder/Bind_T_BinderOptions.generated.txt
Comment (steveharter) [reply src/libraries/Microsoft.Extensions.Configuration.Binder/tests/SourceGenerationTests/Baselines/OptionsBuilder/Bind_T_BinderOptions.generated.txt:113]: Thanks

File: src/libraries/Microsoft.Extensions.Configuration.Binder/tests/SourceGenerationTests/Baselines/OptionsBuilder/Bind_T_BinderOptions.generated.txt
Comment (steveharter) [reply src/libraries/Microsoft.Extensions.Configuration.Binder/tests/SourceGenerationTests/Baselines/OptionsBuilder/Bind_T_BinderOptions.generated.txt:122]: A newer commit changed the `else` to `else if` so I don't think a conditional assignment works well anymore.

File: src/libraries/Microsoft.Extensions.Configuration.Binder/tests/SourceGenerationTests/Baselines/Collections.generated.txt
Comment (steveharter) [reply src/libraries/Microsoft.Extensions.Configuration.Binder/tests/SourceGenerationTests/Baselines/Collections.generated.txt]: A later commit addresses this; it should be `false` here to match reflection semantics.

File: src/libraries/Microsoft.Extensions.Configuration.Binder/tests/SourceGenerationTests/Baselines/ConfigurationBinder/Bind.generated.txt
Comment (layomia) [inline src/libraries/Microsoft.Extensions.Configuration.Binder/tests/SourceGenerationTests/Baselines/ConfigurationBinder/Bind.generated.txt:48]:
```suggestion
            BindCore(configuration, ref typedObj, defaultValueIfNotFound: false, binderOptions: null);
```

File: src/libraries/Microsoft.Extensions.Configuration.Binder/gen/Emitter/CoreBindingHelpers.cs
Comment (layomia) [inline src/libraries/Microsoft.Extensions.Configuration.Binder/gen/Emitter/CoreBindingHelpers.cs:314]: > // Currently properties don't have an 'else' failure case.

Seems like we should remove the abstraction to avoid these two features from clashing. Not sure that the property binding expectation here will change.

File: src/libraries/Microsoft.Extensions.Configuration.Binder/gen/Emitter/CoreBindingHelpers.cs
Comment (layomia) [inline src/libraries/Microsoft.Extensions.Configuration.Binder/gen/Emitter/CoreBindingHelpers.cs:642]: Seems to me that `InitializationKind.Declaration` and `ValueDefaulting.None` indicate the same scenario, when a new object is being declared. Maybe we could remove `ValueDefaulting`?

File: src/libraries/Microsoft.Extensions.Configuration.Binder/gen/Emitter/CoreBindingHelpers.cs
Comment (layomia) [inline src/libraries/Microsoft.Extensions.Configuration.Binder/gen/Emitter/CoreBindingHelpers.cs:387]: Would `InitializationKind.Declaration` express the same intent?

File: src/libraries/Microsoft.Extensions.Configuration.Binder/gen/Emitter/CoreBindingHelpers.cs
Comment (steveharter) [reply src/libraries/Microsoft.Extensions.Configuration.Binder/gen/Emitter/CoreBindingHelpers.cs:642]: The only place with `ValueDefaulting.CallSetter` is for complex types (objects) which also has  `InitializationKind.Declaration`. However, arrays and enumerables also have  `InitializationKind.Declaration` but should not have `ValueDefaulting.CallSetter`.

File: src/libraries/Microsoft.Extensions.Configuration.Binder/gen/Emitter/CoreBindingHelpers.cs
Comment (steveharter) [reply src/libraries/Microsoft.Extensions.Configuration.Binder/gen/Emitter/CoreBindingHelpers.cs:387]: Sure.

File: src/libraries/Microsoft.Extensions.Configuration.Binder/gen/Emitter/CoreBindingHelpers.cs
Comment (layomia) [reply src/libraries/Microsoft.Extensions.Configuration.Binder/gen/Emitter/CoreBindingHelpers.cs:642]: Seems like something we can simplify by adding intermediary logic/methods before we get to the core binding methods. I wouldn't block on this though.

Generally I'm trying to tease out the decisions made in the core binding methods so that it's to understand (https://github.com/dotnet/runtime/pull/91717#discussion_r1320459650).

File: src/libraries/Microsoft.Extensions.Configuration.Binder/tests/SourceGenerationTests/Baselines/ConfigurationBinder/Bind.generated.txt
Comment (steveharter) [reply src/libraries/Microsoft.Extensions.Configuration.Binder/tests/SourceGenerationTests/Baselines/ConfigurationBinder/Bind.generated.txt:48]: OK; originally I was assuming we wanted terse generated code, but I see that we use `binderOptions: null`

File: src/libraries/Microsoft.Extensions.Configuration.Binder/gen/Emitter/CoreBindingHelpers.cs
Comment (steveharter) [reply src/libraries/Microsoft.Extensions.Configuration.Binder/gen/Emitter/CoreBindingHelpers.cs:636]: The reason for the difference is because of the non-intuitive reflection semantics around not calling setter for objects added to arrays.

I could look at adding support for collections\enumerable; it would be a small breaking change for reflection, but that would mean modifying the reflection code.

File: src/libraries/Microsoft.Extensions.Configuration.Binder/gen/Emitter/CoreBindingHelpers.cs
Comment (steveharter) [reply src/libraries/Microsoft.Extensions.Configuration.Binder/gen/Emitter/CoreBindingHelpers.cs]: I moved the ErrorOnFailedBindingProperty from the base class to the parameter class, and removed this assert.

File: src/libraries/Microsoft.Extensions.Configuration.Binder/tests/SourceGenerationTests/Baselines/ConfigurationBinder/Bind.generated.txt
Comment (layomia) [reply src/libraries/Microsoft.Extensions.Configuration.Binder/tests/SourceGenerationTests/Baselines/ConfigurationBinder/Bind.generated.txt]: We do want terse code, but generally prioritize readability.

File: src/libraries/Microsoft.Extensions.Configuration.Binder/gen/Emitter/CoreBindingHelpers.cs
Comment (layomia) [reply src/libraries/Microsoft.Extensions.Configuration.Binder/gen/Emitter/CoreBindingHelpers.cs:636]: > I could look at adding support for collections\enumerable; it would be a small breaking change for reflection, but that would mean modifying the reflection code.

We should definitely avoid this, especially not for 8. The binder / user expectations are very sensitive to any breaking changes, as @tarekgh would probably echo.

Comment (layomia) [reply src/libraries/Microsoft.Extensions.Configuration.Binder/gen/Emitter/CoreBindingHelpers.cs:636]: The approach you have now looks good to me.

Comment (tarekgh) [reply src/libraries/Microsoft.Extensions.Configuration.Binder/gen/Emitter/CoreBindingHelpers.cs:636]: Yes, please try avoiding introducing any breaking change in for the reflection. Any slight change gets noticed and causes us servicing it. At least try to avoid it in the current stage of the project.

**Review by layomia — APPROVED** [review]
Thanks!

**Comment by steveharter** [comment]
Need to get in at least this before backport merge can be successful: https://github.com/dotnet/runtime/pull/91717

**Pull request merged**
